### PR TITLE
Working draft of relstorage support for postgresql

### LINKED ISSUE #124: RelStorage support
Add possibility to use RelStorage instead of ZEO via **environment variables** or **command**

### PR DESCRIPTION
**Pull request by pnicolli**
This is my go at adding RelStorage support.
Fixes #124 

A few notes:

- This only supports postgresql for now, and only on the 5.2.2 debian image.
- It felt right to add an env var for each RelStorage options, but I kept a single comma-separated env var for adapter-specific options, because they differ and I didn't think doing several specific ``if-else``s was a good idea. Thoughts about this?
- Should we separate the ``relstorage_conf`` method to a different file?
- Before going live with this I would probably test at least adding a mysql driver together with the postgresql one and seeing that it works, otherwise it would make no sense to only support postgresql, imho.
- The other option would have been to let the user specify the adapter by adding the RelStorage egg via the ``ADDONS`` env var, but then that would mean adding several build libraries (e.g. ``gcc`` and ``libpq-dev``) to the ``runDeps``, and I wasn't sure I liked it.
- I set a default value for the ``blob-dir`` option. From what I read in the RelStorage documentation, I understood that it is needed in order to handle blobs. Does that mean blobs are held inside postgres and the ``blob-dir`` is just a cache? Because that's what I wanted to achieve :)
- Yes, I will also need to upgrade the README according to the new env vars somehow. And they are a lot 😅 

I built a version of this image and it's available on docker hub as ``pnicolli/plone:5.2.2-rt3``, so it's ready for you to try.

Steps to try the image:

```
$ docker run -d --rm -e POSTGRES_USER=zope -e POSTGRES_PASSWORD=zope -e POSTGRES_DB=zope --name pg postgres
$ docker run --rm -e RELSTORAGE_ADAPTER_OPTIONS="type postgresql,host pg,dbname zope,user zope,password zope" --link pg -p 8080:8080 --name plone pnicolli/plone:5.2.2-rt3
```

### CODE REVIEW

**Comment by avoinea** [comment]
@pnicolli Wow, THANK YOU!

As this will grow and will need also MySQL support (and why not Oracle) at some point, I would go by splitting it by tags. 

Thus, as now we have:
plone:5
plone:5-alpine
plone:5-python36
...

We can have:
plone:5-relstorage (Default PostgreSQL)
plone:5-relstorage-mysql
plone:5-relstorage-oracle
...

@jensens, @svx Suggestions?!?

**Comment by jensens** [comment]
Love that! I am not sure if we really should go splitting it so much in detail. 

What prevents us to ship the image with all adapter options installed: `pip install "RelStorage[mysql,postgresql,oracle]`?

Ok, it might get slightly larger and we need to `apt install default-libmysqlclient-dev` - but for the convinienve of one "it just works" image?

File: 5.2/5.2.2/debian/buildout.cfg [hunk 1]
```diff
@@ -12,6 +12,10 @@ parts +=
   mrbob
   plonesite
 
+eggs +=
+  psycopg2
+  RelStorage
```
Comment (jensens) [inline 5.2/5.2.2/debian/buildout.cfg:17]:
```suggestion
  RelStorage[mysql,postgresql,oracle]
```

File: 5.2/5.2.2/debian/Dockerfile [hunk 2]
```diff
@@ -21,8 +21,8 @@ RUN useradd --system -m -d /plone -U -u 500 plone \
 
 COPY buildout.cfg /plone/instance/
 
-RUN buildDeps="dpkg-dev gcc libbz2-dev libc6-dev libffi-dev libjpeg62-turbo-dev libopenjp2-7-dev libpcre3-dev libssl-dev libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev" \
- && runDeps="git gosu libjpeg62 libopenjp2-7 libtiff5 libxml2 libxslt1.1 lynx netcat poppler-utils rsync wv" \
+RUN buildDeps="dpkg-dev gcc libbz2-dev libc6-dev libffi-dev libjpeg62-turbo-dev libopenjp2-7-dev libpcre3-dev libpq-dev libssl-dev libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev" \
```
Comment (jensens) [inline 5.2/5.2.2/debian/Dockerfile:24]:
```suggestion
RUN buildDeps="default-libmysqlclient-dev dpkg-dev gcc libbz2-dev libc6-dev libffi-dev libjpeg62-turbo-dev libopenjp2-7-dev libpcre3-dev libpq-dev libssl-dev libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev" \
```

**Comment by avoinea** [comment]
@jensens I'm also the fan of one "it just works" image as it will save us from maintaining the matrix of multiple variations.

If we go on this path, then it should also work on alpine :smile: 

Or we can keep the alpine and a debian slim version and have another version full batteries included.

File: 5.2/5.2.2/debian/buildout.cfg (hunk 1 above)
Comment (pnicolli) [reply 5.2/5.2.2/debian/buildout.cfg:17]: Oh, I didn't know I could add more than one extra this way, thanks :)

**Comment by pnicolli** [comment]
At first, I also thought about separating tags based on relstorage adapter support, but I'd prefer having an all-in-one solution, so I went for the "just install RelStorage and debian dependencies right away" path :) 
Yes, I think alpine should work as well, I will try that.

**Comment by pnicolli** [comment]
I might need help for mysql, I have never tried it. I added the packages suggested by @jensens and published an updated version of the docker image ``pnicolli/plone:5.2.2-rt3`` and I am testing it with this docker-compose configuration:

```yaml
version: '3'

services:
  mysql:
    image: mysql
    environment:
      MYSQL_ROOT_PASSWORD: mysql
      MYSQL_DATABASE: zope
      MYSQL_USER: zope
      MYSQL_PASSWORD: zope
  postgres:
    image: postgres
    environment:
      POSTGRES_USER: zope
      POSTGRES_PASSWORD: zope
      POSTGRES_DB: zope
  plone:
    image: pnicolli/plone:5.2.2-rt3
    environment:
      #RELSTORAGE_ADAPTER_OPTIONS: "type postgresql,host postgres,dbname zope,user zope,password zope"
      RELSTORAGE_ADAPTER_OPTIONS: "type mysql,host mysql,db zope,user zope,passwd zope"
    ports:
      - 8080:8080
```

But it's not working, I am getting the following error while starting the plone instance:

```
plone_1     | Traceback (most recent call last):
plone_1     |   File "/plone/buildout-cache/eggs/mysqlclient-2.0.2-py3.8-linux-x86_64.egg/MySQLdb/__init__.py", line 18, in <module>
plone_1     |     from . import _mysql
plone_1     | ImportError: libmariadb.so.3: cannot open shared object file: No such file or directory

plone_1     | During handling of the above exception, another exception occurred:
plone_1     | 
plone_1     | Traceback (most recent call last):
[...]
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/config.py", line 42, in open
plone_1     |     adapter = config.adapter.create(options)
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/config.py", line 75, in create
plone_1     |     return MySQLAdapter(options=options, **params)
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/mysql/adapter.py", line 109, in __init__
plone_1     |     super(MySQLAdapter, self).__init__(options)
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/adapter.py", line 64, in __init__
plone_1     |     self.driver = driver = self._select_driver()
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/adapter.py", line 95, in _select_driver
plone_1     |     return _select_driver(
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/drivers.py", line 44, in _select_driver
plone_1     |     driver = _select_driver_by_name(options.driver, driver_options)
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/drivers.py", line 58, in _select_driver_by_name
plone_1     |     result = factory()
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/drivers.py", line 342, in __call__
plone_1     |     return self.driver_type()
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/mysql/drivers/__init__.py", line 145, in __init__
plone_1     |     super(AbstractMySQLDriver, self).__init__()
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/drivers.py", line 147, in __init__
plone_1     |     self.driver_module = mod = self.get_driver_module()
plone_1     |   File "/plone/buildout-cache/eggs/RelStorage-3.4.0-py3.8-linux-x86_64.egg/relstorage/adapters/drivers.py", line 176, in get_driver_module
plone_1     |     return importlib.import_module(self.MODULE_NAME)
plone_1     |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
plone_1     |     return _bootstrap._gcd_import(name[level:], package, level)
plone_1     |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
plone_1     |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
plone_1     |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
plone_1     |   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
plone_1     |   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
plone_1     |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
plone_1     |   File "/plone/buildout-cache/eggs/mysqlclient-2.0.2-py3.8-linux-x86_64.egg/MySQLdb/__init__.py", line 24, in <module>
plone_1     |     version_info, _mysql.version_info, _mysql.__file__
plone_1     | NameError: name '_mysql' is not defined
```

Bonus question: I actually have no idea how to test Oracle :) I am open to suggestions.

**Comment by jensens** [comment]
> Bonus question: I actually have no idea how to test Oracle :) I am open to suggestions.

Looks like there are ways... (never tried).
https://dev.to/amusarra/how-to-setup-docker-container-oracle-database-19c-for-liferay-development-environment-f0j

**Comment by pnicolli** [comment]
Found a way to make MySQL work. Image ``pnicolli/plone:5.2.2-rt4`` can be used for testing this.

Caveats: In order to make the ``mysql:8`` image work, you need to explicitly set the command to ``--default-authentication-plugin=mysql_native_password``, because of this problem: https://github.com/PyMySQL/mysqlclient/issues/442. I haven't tested ``mysql:5``.

This is now ready for review.

**Review by avoinea — APPROVED** [review]
LGTM

**Pull request merged**